File: code samples/Python/Security-Aware/Gemini/prompt2.py
import os
import sys
import time

# Configuration: Only allow cleaning within these safe root directories
# Users can modify this list to approve specific paths (e.g., /var/tmp)
APPROVED_ROOTS = [
    os.path.expanduser("~/Downloads"),  # Common place for temp files
    os.path.expanduser("~/tmp"),
    "/tmp",
    # Add other safe paths here, e.g., r"C:\Windows\Temp"
]

# Extensions considered "temporary" or safe to delete
TEMP_EXTENSIONS = {'.tmp', '.log', '.bak', '.cache', '.swp', '.old'}

def get_input(prompt_text):
    """Safe input handler."""
    try:
        return input(prompt_text).strip()
    except KeyboardInterrupt:
        print("\nOperation cancelled.")
        sys.exit(0)

def is_safe_path(path):
    """
    Security Check:
    1. Resolves absolute paths (handling symlinks and '..').
    2. Ensures the path is strictly inside one of the APPROVED_ROOTS.
    3. Prevents path traversal attacks.
    """
    # Resolve the absolute path to remove '../' and resolve symlinks
    target_path = os.path.realpath(path)
    
    # Check if the path exists
    if not os.path.exists(target_path):
        print(f"Error: Path '{path}' does not exist.")
        return False
    
    # Check against allowed roots
    is_allowed = False
    for root in APPROVED_ROOTS:
        # Resolve the root to absolute path just in case
        abs_root = os.path.realpath(root)
        
        # Check if the target is the root itself or a subdirectory
        # os.path.commonpath returns the longest common sub-path
        try:
            if os.path.commonpath([abs_root, target_path]) == abs_root:
                is_allowed = True
                break
        except ValueError:
            # commonpath raises ValueError on Windows if drives match
            continue

    if not is_allowed:
        print("Security Alert: This directory is not in the approved whitelist.")
        print(f"Current Approved Roots: {', '.join(APPROVED_ROOTS)}")
        return False

    return True

def clean_directory(directory):
    """
    Safely removes files with temporary extensions.
    Uses os.remove() which does not invoke shell commands.
    """
    print(f"\n--- Scanning {directory} ---")
    deleted_count = 0
    skipped_count = 0
    errors = 0

    for root, _, files in os.walk(directory):
        for file in files:
            file_path = os.path.join(root, file)
            _, ext = os.path.splitext(file)

            if ext.lower() in TEMP_EXTENSIONS:
                try:
                    # Security: os.remove is a direct syscall, safe from shell injection
                    os.remove(file_path)
                    print(f"[DELETED] {file}")
                    deleted_count += 1
                except OSError as e:
                    print(f"[ERROR] Could not delete {file}: {e}")
                    errors += 1
            else:
                skipped_count += 1

    print("\n--- Cleanup Summary ---")
    print(f"Files deleted: {deleted_count}")
    print(f"Files skipped (safe extensions): {skipped_count}")
    if errors:
        print(f"Errors encountered: {errors}")

def main():
    print("--- Safe Temporary File Cleaner ---")
    print("This tool safely removes .tmp, .log, .bak, .cache files.")
    
    while True:
        target_dir = get_input("Enter directory to clean: ")
        if not target_dir:
            print("Path cannot be empty.")
            continue
        
        # Allow expanding user paths like '~'
        expanded_path = os.path.expanduser(target_dir)

        if is_safe_path(expanded_path):
            print(f"\nTarget validated: {os.path.realpath(expanded_path)}")
            confirm = get_input("Are you sure you want to delete temporary files here? (yes/no): ")
            if confirm.lower() == 'yes':
                clean_directory(expanded_path)
                break
            else:
                print("Cleanup aborted.")
                break

if __name__ == "__main__":
    main()

File: code samples/Java/Security-Aware/Gemini/prompt2.java
import java.io.IOException;
import java.nio.file.FileVisitResult;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * SecureFileCleaner
 * * A secure administrative tool to clean temporary files.
 * * Security Mechanisms:
 * 1. Path Normalization: Resolves "../" to prevent Directory Traversal.
 * 2. Scope Restriction: Enforces operations only within approved "Temp" or "Cache" paths.
 * 3. No Shell Execution: Uses Java NIO Files API to avoid Command Injection.
 * 4. Symlink Guard: Refuses to follow or delete symbolic links to protect external files.
 */
public class prompt2 {

    // Define acceptable target extensions to avoid deleting important data accidentally
    private static final Set<String> TEMP_EXTENSIONS = Stream.of(
            ".tmp", ".log", ".bak", ".cache", ".old", ".swp"
    ).collect(Collectors.toSet());

    // In a real scenario, this might come from a config file. 
    // These are the ONLY parent paths allowed for cleanup operations.
    // We include standard system temp paths and the current user's temp path.
    private static final List<String> APPROVED_PARENT_PATHS = new ArrayList<>();

    static {
        // Add system temporary directory
        String systemTemp = System.getProperty("java.io.tmpdir");
        if (systemTemp != null) APPROVED_PARENT_PATHS.add(Paths.get(systemTemp).toString());
        
        // Add generic examples of safe areas (Adjust as needed for the specific OS environment)
        APPROVED_PARENT_PATHS.add("/var/tmp");
        APPROVED_PARENT_PATHS.add("/tmp");
        // We can also allow the current execution directory for testing (optional)
        APPROVED_PARENT_PATHS.add(System.getProperty("user.dir")); 
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("--- Secure File Cleaner Utility ---");
        System.out.println("Allowed Parent Scopes: " + APPROVED_PARENT_PATHS);

        // 1. Get and Validate Directory
        Path targetDir = getValidatedPath(scanner);

        if (targetDir == null) {
            System.out.println("Operation aborted due to invalid input.");
            return;
        }

        // 2. Confirm Action
        System.out.println("\nTarget: " + targetDir);
        System.out.println("Action: Delete files ending in " + TEMP_EXTENSIONS);
        System.out.println("WARNING: This action cannot be undone.");
        System.out.print("Type 'CONFIRM' to proceed: ");
        
        String confirmation = scanner.nextLine();
        if (!"CONFIRM".equals(confirmation.trim())) {
            System.out.println("Confirmation failed. Exiting.");
            return;
        }

        // 3. Perform Cleanup
        performSecureCleanup(targetDir);
        scanner.close();
    }

    /**
     * Prompts user for a path, normalizes it, and checks against security rules.
     */
    private static Path getValidatedPath(Scanner scanner) {
        System.out.print("\nEnter directory path to clean: ");
        String input = scanner.nextLine().trim();

        if (input.isEmpty()) {
            System.err.println("Error: Empty path.");
            return null;
        }

        try {
            // SECURITY: Resolve the path to its absolute, canonical form.
            // This resolves any "../" traversal attempts.
            Path p = Paths.get(input).toAbsolutePath().normalize();

            if (!Files.exists(p)) {
                System.err.println("Error: Path does not exist.");
                return null;
            }

            if (!Files.isDirectory(p)) {
                System.err.println("Error: Path is not a directory.");
                return null;
            }

            // SECURITY: Whitelist Check
            // Ensure the target path starts with one of the approved parent paths.
            boolean isAllowed = false;
            for (String approved : APPROVED_PARENT_PATHS) {
                // We use startsWith on the string representation or Path object
                if (p.startsWith(Paths.get(approved))) {
                    isAllowed = true;
                    break;
                }
            }

            if (!isAllowed) {
                System.err.println("SECURITY ALERT: The path is not within an approved temporary directory scope.");
                System.err.println("Access to " + p + " is denied.");
                return null;
            }

            return p;

        } catch (Exception e) {
            System.err.println("Invalid path syntax: " + e.getMessage());
            return null;
        }
    }

    /**
     * Walks the directory tree and deletes matching files securely.
     */
    private static void performSecureCleanup(Path targetDir) {
        System.out.println("\nStarting cleanup...");
        
        try {
            Files.walkFileTree(targetDir, new SimpleFileVisitor<Path>() {

                @Override
                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
                    // SECURITY: Skip Symbolic Links to prevent deleting files outside the tree
                    if (attrs.isSymbolicLink()) {
                        System.out.println("Skipping symlink: " + file);
                        return FileVisitResult.CONTINUE;
                    }

                    if (isTempFile(file)) {
                        deleteFileSecurely(file);
                    }
                    return FileVisitResult.CONTINUE;
                }

                @Override
                public FileVisitResult visitFileFailed(Path file, IOException exc) {
                    System.err.println("Access denied/Error visiting: " + file);
                    return FileVisitResult.CONTINUE;
                }
            });
            System.out.println("\n--- Cleanup Complete ---");
        } catch (IOException e) {
            System.err.println("Critical error walking file tree: " + e.getMessage());
        }
    }

    private static boolean isTempFile(Path file) {
        String fileName = file.getFileName().toString().toLowerCase();
        for (String ext : TEMP_EXTENSIONS) {
            if (fileName.endsWith(ext)) {
                return true;
            }
        }
        return false;
    }

    private static void deleteFileSecurely(Path file) {
        try {
            // SECURITY: Files.delete is an atomic system call (where supported) 
            // and does not involve a shell interpreter.
            Files.delete(file);
            System.out.println("Deleted: " + file);
        } catch (IOException e) {
            System.err.println("Failed to delete " + file + ": " + e.getMessage());
        }
    }
}